woocommerce/google-listings-and-ads
# Add activation from source tracks event

### Changes proposed in this Pull Request:

Closes #990.

This is a proof-of-concept PR for retrieving source/campaign values from a deep-linked campaign when the plugin is activated.

It requires a deep-link to the `Add Plugins` page, with pre-established source parameters:
```
…/wp-admin/plugin-install.php?s=google+listings+%26+ads&tab=search&type=term&traffic_source=online_ad
```

The exact name of `traffic_source` parameter can be set in the event class, which also supports multiple params (e.g., `&campaign_id=123&campaign_medium=facebook`).


**Notes**
- Tracks events are currently registered in `init`, which hasn't been reached during activation, so executing on another hook also won't work as the listener hasn't been registered. The workaround is to invoke the event's tracking method manually after registering all the plugin services in the plugin `activate` method.
- The supported UTM query parameters are `utm_source`, `utm_medium`, `utm_campaign`, `utm_term` and  `utm_content`


### Detailed test instructions:

1. Deactivate the GLA plugin.
2. Use [this handy snippet](https://github.com/woocommerce/google-listings-and-ads/pull/725#pullrequestreview-675174014) from @mikkamp to log tracks requests:
    ```php
     add_filter(
         'pre_http_request',
         function( $pre, $args, $url ) {
              if ( preg_match( '#^https://pixel\.wp\.com#', $url ) ) {
                   do_action( 'woocommerce_gla_debug_message', $url, 'Tracks URL' );
              }
              return $pre;
         },
         10,
         3
    );
    ```
3. Make sure your WooCommerce store has tracking enabled WooCommerce > Settings > Advanced > WooCommerce.com > Enable Tracking (`/wp-admin/admin.php?page=wc-settings&tab=advanced&section=woocommerce_com`)
4. Visit this link from a hypothetical deep link campaign (where the advertiser already knows your domain): 
     `/wp-admin/plugin-install.php?s=google+listings+%26+ads&tab=search&type=term&traffic_source=online_ad&traffic_type=mobile`
    (Note that the `traffic_` params are currently hard-coded, but could be changed to something else, or made to dynamically include everything except `s`, `tab`, and `type`.
5. You should see GLA as the first option, with a gray button to `Activate`. 
    (Note that, if GLA _weren't_ installed, you would first see the `Install` button, which actually runs an async install process and thus doesn't redirect the user. You can confirm this by `Install`ing one of the other plugins that shows up in the search results).
6. Click the `Activate` button, and then check in your log for the corresponding Tracks requests (WooCommerce > Status > Logs – `/wp-admin/admin.php?page=wc-status&tab=logs`)
    - Look for the event `wcadmin_gla_activated_from_source ` and confirm that the tracking URL contains the `traffic_source` and `traffic_type` params


### Changelog entry
